contrib/platform/src/com/sun/jna/platform/win32/Cfgmgr32Util.java

```diff
@@ -96,4 +96,44 @@ public static String CM_Get_Device_ID(int devInst) throws Cfgmgr32Exception {
             return buffer.getWideString(0);
         }
     }
+
+    /**
+     * Utility method to call Cfgmgr32's CM_Get_DevNode_Registry_Property that
+     * allocates the required memory for the Buffer parameter, and and returns the
+     * received string.
+     *
+     * @param devInst
+     *            Caller-supplied device instance handle that is bound to the local
+     *            machine.
+     * @param ulProperty
+     *            A {@code CM_DRP_}-prefixed constant value that identifies the
+     *            device property to be obtained from the registry. These constants
+     *            are defined in Cfgmgr32.h.
+     * @return The device instance ID string.
+     * @throws Cfgmgr32Exception
+     */
+    public static String CM_Get_DevNode_Registry_Property(int devInst, int ulProperty) throws Cfgmgr32Exception {
```

Comment:
I don't think this is valid. To my knowledge the registry supports more datatypes, than just string. If I read the signature correctly, the third parameter for the call to `CM_Get_DevNode_Registry_Property` `pulRegDataType` needs to be supplied and checked. Based on that type, the right datatype can be determined:

```
#define REG_NONE                    ( 0 )   // No value type
#define REG_SZ                      ( 1 )   // Unicode nul terminated string
#define REG_EXPAND_SZ               ( 2 )   // Unicode nul terminated string
                                            // (with environment variable references)
#define REG_BINARY                  ( 3 )   // Free form binary
#define REG_DWORD                   ( 4 )   // 32-bit number
#define REG_DWORD_LITTLE_ENDIAN     ( 4 )   // 32-bit number (same as REG_DWORD)
#define REG_DWORD_BIG_ENDIAN        ( 5 )   // 32-bit number
#define REG_LINK                    ( 6 )   // Symbolic Link (unicode)
#define REG_MULTI_SZ                ( 7 )   // Multiple Unicode strings
#define REG_RESOURCE_LIST           ( 8 )   // Resource list in the resource map
#define REG_FULL_RESOURCE_DESCRIPTOR ( 9 )  // Resource list in the hardware description
#define REG_RESOURCE_REQUIREMENTS_LIST ( 10 )
#define REG_QWORD                   ( 11 )  // 64-bit number
#define REG_QWORD_LITTLE_ENDIAN     ( 11 )  // 64-bit number (same as REG_QWORD)
```

The corresponding values are already bound in `WinNT.java`: 

https://github.com/java-native-access/jna/blob/58bc0a00b0789205dca65fe0299e27c5f2f57572/contrib/platform/src/com/sun/jna/platform/win32/WinNT.java#L1086-L1154

Reply:
Fair point, but a few counter points:
 - The docs do state this argument is optional
 - The types returned are well defined (see comments in the `CM_DRP_*` constant mappings) 
 - In my particular use case I'm only ever fetching the string (`REG_SZ`) properties (Desc, friendly name, driver, class, and service).  Which this util provides for. Perhaps its name should make that clear, such as `CM_Get_DevNode_Registry_Property_Reg_Sz`.
   - The CM_DRP properties do include `REG_MULTI_SZ` and `DWORD` values, so one option might be to create another util method that returns a string array, perhaps a third that returns an `int`, and name the methods `CM_Get_DevNode_Registry_Property_Reg_Multi_Sz`, and `CM_Get_DevNode_Registry_Property_Reg_Dword`.  
 - Or I could just delete the util method entirely.

Reply:
I added some new methods as proposed above. Let me know if you concur with this approach and I'll add to the unit tests.

Reply:
What would you think about a utility method, that returns `Object`, you could then read the type of the registry key and convert after reading to the correct java type (Integer, String[], String, ...)?

Reply:
> a utility method, that returns `Object`

The more complicated we make it for the user, the less utility the utility method has.  The existing two-method calls aren't that hard to use.

Now if there were a `RegistryProperty` object with getters like `getString()`, `getStringArray()`, and `getInt()` that might be easier for the user. 

I just had a look at `Advapi32Util` which essentially queries the same type of objects, and it has methods like `registryGetStringValue`, `registryGetStringArray`, and `registryGetIntValue` which I think matches what I'm doing (with names representing the return vs. the registry type)

Reply:
I think more methods make it harder to find the right one, but we can agree to disagree here. The core point is: You don't know the type, until you query it and you need to be prepared, that it can explode in your face (either with an IllegalArgumentException in your case, or an ClassCastException).

For the "make it easy for people" argumet: We are talking C level access here - people can shoot themselves in the foot. HARD. My sympathy with people not being able to do a basic java cast is zero, as they will screw up memory management 100% of the time.

Rant aside, ok, align the names and throws Exceptions with `Advapi32Util` and I'll shut up.

Reply:
Still thinking about this, and I'm leaning toward single method/return type.  Ultimately we're querying a buffer of memory so we just need a class which extends `PointerType` (so a user who knows what they are doing can access the memory buffer directly) which stores the type privately and has type-checking utility methods for the string, string array, and int.

Reply:
I don't think this will work. A pointer does not hold enough information. Consider a Memory object holding an integer. If that Integer is 0xFFFFFFFF (or something else, that fills every byte), calling getString or getWideString on the memory object will cause an out of bounds read. Compare this with a java Object:

a) it contains type information, that you can check (instanceof)
b) can be savely casted and will raise a catchable exception if type does not match

Reply:
I think my latest update addresses the concerns above.